// himalaya/lib/formats/v0.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }(); /*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           This format adheres to the v0 ASP spec.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         */


exports.default = format;
exports.capitialize = capitialize;
exports.camelCase = camelCase;
exports.castValue = castValue;
exports.unquote = unquote;
exports.splitHead = splitHead;
exports.formatAttributes = formatAttributes;
exports.formatStyles = formatStyles;

var _compat = require('../compat');

function format(nodes) {
  return nodes.map(function (node) {
    var type = capitialize(node.type);
    if (type === 'Element') {
      var tagName = node.tagName.toLowerCase();
      var attributes = formatAttributes(node.attributes);
      var children = format(node.children);
      return { type: type, tagName: tagName, attributes: attributes, children: children };
    }

    return { type: type, content: node.content };
  });
}

function capitialize(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function camelCase(str) {
  return str.split('-').reduce(function (str, word) {
    return str + word.charAt(0).toUpperCase() + word.slice(1);
  });
}

function castValue(str) {
  if (typeof str !== 'string') return str;
  if (str === '') return str;
  var num = +str;
  if (!isNaN(num)) return num;
  return str;
}

function unquote(str) {
  var car = str.charAt(0);
  var end = str.length - 1;
  if (car === '"' || car === "'" && car === str.charAt(end)) {
    return str.slice(1, end);
  }
  return str;
}

function splitHead(str, sep) {
  var idx = str.indexOf(sep);
  if (idx === -1) return [str];
  return [str.slice(0, idx), str.slice(idx + sep.length)];
}

function formatAttributes(attributes) {
  return attributes.reduce(function (attrs, pair) {
    var _splitHead = splitHead(pair.trim(), '='),
        _splitHead2 = _slicedToArray(_splitHead, 2),
        key = _splitHead2[0],
        value = _splitHead2[1];

    value = value ? unquote(value) : key;
    if (key === 'class') {
      attrs.className = value.split(' ');
    } else if (key === 'style') {
      attrs.style = formatStyles(value);
    } else if ((0, _compat.startsWith)(key, 'data-')) {
      attrs.dataset = attrs.dataset || {};
      var prop = camelCase(key.slice(5));
      attrs.dataset[prop] = castValue(value);
    } else {
      attrs[camelCase(key)] = castValue(value);
    }
    return attrs;
  }, {});
}

function formatStyles(str) {
  return str.trim().split(';').map(function (rule) {
    return rule.trim().split(':');
  }).reduce(function (styles, keyValue) {
    var _keyValue = _slicedToArray(keyValue, 2),
        rawKey = _keyValue[0],
        rawValue = _keyValue[1];

    if (rawValue) {
      var key = camelCase(rawKey.trim());
      var value = castValue(rawValue.trim());
      styles[key] = value;
    }
    return styles;
  }, {});
}
